mu4/notation/internal/notationnoteinput.cpp

```diff
@@ -0,0 +1,283 @@
+        }
+        is.setString(strg);
+        break;
+    }
+    case Ms::StaffGroup::PERCUSSION:
+        break;
+    }
+
+    notifyAboutStateChanged();
+}
+
+void NotationNoteInput::endNoteInput()
+{
+    Ms::InputState& is = score()->inputState();
+    is.setNoteEntryMode(false);
+    if (is.slur()) {
+        const std::vector<Ms::SpannerSegment*>& el = is.slur()->spannerSegments();
+        if (!el.empty()) {
+            el.front()->setSelected(false);
+        }
+        is.setSlur(0);
+    }
+
+    notifyAboutStateChanged();
+}
+
+void NotationNoteInput::toggleNoteInputMethod(NoteInputMethod method)
+{
+    Ms::InputState& inputState = score()->inputState();
+    inputState.setNoteEntryMethod(method);
+
+    notifyAboutStateChanged();
+}
+
+void NotationNoteInput::addNote(NoteName noteName, NoteAddingMode addingMode)
+{
+    if (!isNoteInputMode()) {
+        startNoteInput();
+    }
+
+    Ms::EditData editData;
+    editData.view = m_scoreCallbacks;
+
+    m_undoStack->prepareChanges();
+    int inote = static_cast<int>(noteName);
+    bool addToUpOnCurrentChord = addingMode == NoteAddingMode::CurrentChord;
+    bool insertNewChord = addingMode == NoteAddingMode::InsertChord;
+    score()->cmdAddPitch(editData, inote, addToUpOnCurrentChord, insertNewChord);
+    m_undoStack->commitChanges();
+
+    notifyAboutStateChanged();
+}
+
+void NotationNoteInput::padNote(const Pad& pad)
+{
+    Ms::EditData ed;
+    ed.view = m_scoreCallbacks;
+
+    m_undoStack->prepareChanges();
+    score()->padToggle(pad, ed);
+    m_undoStack->commitChanges();
+
+    notifyAboutStateChanged();
+}
+
+void NotationNoteInput::putNote(const QPointF& pos, bool replace, bool insert)
+{
+    m_undoStack->prepareChanges();
+    score()->putNote(pos, replace, insert);
+    m_undoStack->commitChanges();
+
+    m_noteAdded.notify();
+}
+
+void NotationNoteInput::toogleAccidental(AccidentalType accidentalType)
+{
+    Ms::EditData editData;
+    editData.view = m_scoreCallbacks;
+
+    score()->toggleAccidental(accidentalType, editData);
+
+    notifyAboutStateChanged();
+}
+
+void NotationNoteInput::setCurrentVoiceIndex(int voiceIndex)
+{
+    if (!isVoiceIndexValid(voiceIndex)) {
+        return;
+    }
+
+    Ms::InputState& inputState = score()->inputState();
+    int track = (inputState.track() / VOICES) * VOICES + voiceIndex;
+
+    inputState.setTrack(track);
+
+    if (inputState.segment()) {
+        Ms::Segment* segment = inputState.segment()->measure()->first(Ms::SegmentType::ChordRest);
+        inputState.setSegment(segment);
+    }
+
+    notifyAboutStateChanged();
+}
+
+void NotationNoteInput::addSlur(Ms::Slur* slur)
+{
+    Ms::InputState& inputState = score()->inputState();
+    inputState.setSlur(slur);
+
+    std::vector<Ms::SpannerSegment*> slurSpannerSegments = slur->spannerSegments();
+    if (!slurSpannerSegments.empty()) {
+        slurSpannerSegments.back()->setSelected(true);
+    }
+}
+
+void NotationNoteInput::resetSlur()
```

Comment:
Can it be implemented as

setSlur(nullptr)

?

Reply:
fixed